akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -1376,6 +1376,66 @@ private[stream] object Collect {
     }
 }
 
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] final case class MapAsyncUnordered1[In, Out](f: In => Future[Out])
+    extends GraphStage[FlowShape[In, Out]] {
+  private val in = Inlet[In]("MapAsyncUnordered.in")
+  private val out = Outlet[Out]("MapAsyncUnordered.out")
+  override val shape: FlowShape[In, Out] = FlowShape(in, out)
+  override def initialAttributes: Attributes = DefaultAttributes.mapAsyncUnordered and SourceLocation.forLambda(f)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic with InHandler with OutHandler =
+    new GraphStageLogic(shape) with InHandler with OutHandler {
+      private lazy val decider = inheritedAttributes.mandatoryAttribute[SupervisionStrategy].decider
+      private val invokeFutureCB: Try[Out] => Unit = getAsyncCallback(handle).invoke
+      private var inFlight: Boolean = false
+      override def onPush(): Unit = {
+        try {
+          val future = f(grab(in))
+          future.value match {
+            case None =>
+              inFlight = true
+              future.onComplete(invokeFutureCB)(akka.dispatch.ExecutionContexts.parasitic)
+            case Some(value) =>
+              handle(value)
+          }
+        } catch {
+          case NonFatal(ex) =>
+            decider(ex) match {
+              case Supervision.Stop => failStage(ex)
+              case _                => if (isClosed(in)) completeStage() else pull(in)
+            }
+        }
+      }
+
+      private def handle(value: Try[Out]): Unit = {
+        inFlight = false
+        value match {
+          case Success(elem) if elem != null =>
+            push(out, elem)
```

Comment:
It was do a `tryPull(in)` here, but now

Reply:
I think this behavior can be added back, if you think the PR has any change to be merged.

Reply:
@johanandren This is what I mean. Now it only `push`, and then it will be a new `onPull`.  Previously it `pull` and `push` here. Maybe detached is the wrong word.

Anyway, this `push` - `onPull` - `pull` is completely correct.

Reply:
Ah, so the eager pulling in `MapAsyncUnordered` could potentially have a future result already completed when the actual pull comes in compared to this? Like a one element buffer, interesting, maybe that is what actually explains the perf difference then. 

I tried doing the same change to `MapAsyncUnordered` and that makes the benchmark roughly perform the same on my i5/JDK 11 (and the old impl faster on arm64), so maybe that is what we should do instead of a custom impl (checking that it doesn't have any bad effect on the perf of MapAsyncUnordered for non parallelism = 1 cases first ofc)?